File: apps/api/internal/domain/money_test.go
package domain

import (
	"errors"
	"testing"
)

func TestParseMoney(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		input    string
		currency Currency
		want     int64
		wantErr  bool
	}{
		{"CNY integer", "12", CurrencyCNY, 1200, false},
		{"USD one decimal", "12.3", CurrencyUSD, 1230, false},
		{"EUR two decimals", "12.34", CurrencyEUR, 1234, false},
		{"JPY integer", "12", CurrencyJPY, 12, false},
		{"JPY rejects decimal", "12.0", CurrencyJPY, 0, true},
		{"rejects silent rounding", "12.345", CurrencyCNY, 0, true},
		{"rejects negative", "-1.00", CurrencyCNY, 0, true},
		{"rejects exponent", "1e2", CurrencyCNY, 0, true},
		{"rejects spaces", " 1.00", CurrencyCNY, 0, true},
		{"rejects unknown currency", "1", Currency("GBP"), 0, true},
		{"accepts maximum exact JSON integer", "90071992547409.91", CurrencyCNY, MaxSafeMinorUnits, false},
		{"rejects unsafe JSON integer", "90071992547409.92", CurrencyCNY, 0, true},
		{"rejects int64 overflow", "92233720368547759.00", CurrencyCNY, 0, true},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()
			got, err := ParseMoney(test.input, test.currency)
			if test.wantErr {
				if !errors.Is(err, ErrInvalidInput) {
					t.Fatalf("error = %v, want ErrInvalidInput", err)
				}
				return
			}
			if err != nil {
				t.Fatalf("ParseMoney() error = %v", err)
			}
			if got.MinorUnits != test.want || got.Currency != test.currency {
				t.Fatalf("ParseMoney() = %#v, want minor=%d currency=%s", got, test.want, test.currency)
			}
		})
	}

	if err := (Money{MinorUnits: MaxSafeMinorUnits, Currency: CurrencyCNY}).Validate(); err != nil {
		t.Fatalf("max money must be valid: %v", err)
	}
	if err := (Money{MinorUnits: MaxSafeMinorUnits + 1, Currency: CurrencyCNY}).Validate(); !errors.Is(err, ErrInvalidInput) {
		t.Fatalf("unsafe JSON money error = %v", err)
	}
	if err := (Money{MinorUnits: -1, Currency: CurrencyCNY}).Validate(); !errors.Is(err, ErrInvalidInput) {
		t.Fatalf("negative money error = %v", err)
	}
	if err := (Money{MinorUnits: 1, Currency: Currency("GBP")}).Validate(); !errors.Is(err, ErrInvalidInput) {
		t.Fatalf("unsupported currency error = %v", err)
	}
}

File: apps/web/src/features/facts/money.test.ts
import { describe, expect, it } from 'vitest'
import { formatMinorUnits, minorToDecimalInput, parseDecimalToMinor } from './money'

describe('金额十进制与最小单位互转', () => {
  it('按币种精度换算', () => {
    expect(parseDecimalToMinor('123.45', 'CNY')).toEqual({ minor: 12345 })
    expect(parseDecimalToMinor('123', 'CNY')).toEqual({ minor: 12300 })
    expect(parseDecimalToMinor('123.4', 'CNY')).toEqual({ minor: 12340 })
    expect(parseDecimalToMinor('0', 'CNY')).toEqual({ minor: 0 })
    // JPY 精度为 0，整数即最小单位
    expect(parseDecimalToMinor('123', 'JPY')).toEqual({ minor: 123 })
  })

  // 这些写法后端 domain.ParseMoney 一律拒绝；前端多接受一种，用户就会在提交时
  // 收到一个本可以在输入时说清楚的错误。
  it('拒绝后端同样拒绝的写法', () => {
    for (const bad of ['', ' 1', '1 ', '+1', '1,234.00', '1.', '.5', '1.2.3', 'abc', '-1']) {
      expect(parseDecimalToMinor(bad, 'CNY')).toHaveProperty('error')
    }
  })

  it('小数位超过币种精度时拒绝并说明精度', () => {
    expect(parseDecimalToMinor('1.234', 'CNY')).toEqual({ error: 'CNY 小数位不能超过 2 位' })
    expect(parseDecimalToMinor('1.2', 'JPY')).toEqual({ error: 'JPY 不使用小数位' })
  })

  it('不支持的币种不猜测精度', () => {
    expect(parseDecimalToMinor('1.00', 'GBP')).toHaveProperty('error')
  })

  it('往返一致', () => {
    for (const [minor, currency] of [
      [12345, 'CNY'],
      [0, 'CNY'],
      [123, 'JPY'],
      [99999999, 'USD'],
    ] as const) {
      const text = minorToDecimalInput(minor, currency)
      expect(parseDecimalToMinor(text, currency)).toEqual({ minor })
    }
  })

  it('展示格式保持带币种前缀，输入格式不带', () => {
    expect(formatMinorUnits(12345, 'CNY')).toBe('CNY 123.45')
    expect(minorToDecimalInput(12345, 'CNY')).toBe('123.45')
  })
})
